File: LC328.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {

        if(!head or !head->next)
            return head;
        
        ListNode *h1 = head->next, *i = head;
        ListNode *j = h1;

        while(i->next and j->next){
            i->next = j->next;
            if(i->next)
                i = i->next;
            j->next = i->next;
            if(j->next)
                j = j->next;
        }
        i->next = h1;

        return head;
    }
};

File: LC328.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:

        if not head or not head.next:
            return head
        
        i = head;
        j = h1 = head.next

        while i.next and j.next:
            i.next = j.next
            if i.next:
                i = i.next
            j.next = i.next
            if j.next:
                j = j.next
        
        i.next = h1

        return head
